### ISSUE TITLE
scylla-jmx files aren't installed to wrong directory in upgrade from 3.1.2-0.20191127.9d78d848e 

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla-jmx#90]
Version: scylla-jmx-3.2.rc4-20200101.0a404662d68.noarch

### Test jobs
- https://jenkins.scylladb.com/view/scylla-3.2/job/scylla-3.2/job/rolling-upgrade/job/rolling-upgrade-centos7/47/
- https://jenkins.scylladb.com/view/scylla-3.2/job/scylla-3.2/job/rolling-upgrade/job/rolling-upgrade-centos7/48/

### Description:

In upgrade test from 3.1.2-0.20191127.9d78d848e to scylla-jmx-3.2.rc4-20200101.0a404662d68.noarch, scylla-jmx failed to start after upgraded one db node.

### Scylla-JMX error after upgraded one db node

```
| systemd: Started Scylla JMX.
| systemd: Started Run Scylla Housekeeping restart mode.
| systemd: Started Run Scylla Housekeeping daily mode.
| systemd: Failed at step EXEC spawning /usr/lib/scylla/jmx/scylla-jmx: No such file or directory
| systemd: scylla-jmx.service: main process exited, code=exited, status=203/EXEC
| systemd: Unit scylla-jmx.service entered failed state.
| systemd: scylla-jmx.service failed.
```

### SCT key logs:

```
2020-01-01T22:15:26 db-node-b642da38-0-4 !INFO    | scylla: [shard 0] init - Scylla version 3.1.2-0.20191127.9d78d848e initialization completed.

## Upgrade first db node

< t:2020-01-01 22:43:18,615 f:upgrade_test.py l:473  c:sdcm.tester          p:INFO  > Upgrade Node rolling-upgrade-3-2-centos-db-node-b642da38-0-4 begin
< t:2020-01-01 22:43:37,900 f:upgrade_test.py l:111  c:sdcm.tester          p:INFO  > Upgrading a Node

2020-01-01T22:47:01+00:00  rolling-upgrade-3-2-centos-db-node-b642da38-0-4 !INFO    | yum[32369]: Updated: scylla-server-3.2.rc4-0.20200101.c8d5738a482.el7.x86_64
2020-01-01T22:47:01+00:00  rolling-upgrade-3-2-centos-db-node-b642da38-0-4 !INFO    | useradd[32497]: failed adding user 'scylla', exit code: 9
2020-01-01T22:47:02+00:00  rolling-upgrade-3-2-centos-db-node-b642da38-0-4 !INFO    | yum[32369]: Updated: scylla-jmx-3.2.rc4-20200101.0a404662d68.noarch
2020-01-01T22:47:02+00:00  rolling-upgrade-3-2-centos-db-node-b642da38-0-4 !INFO    | yum[32369]: Updated: scylla-3.2.rc4-0.20200101.c8d5738a482.el7.x86_64

< t:2020-01-01 22:47:16,475 f:remote.py       l:122  c:sdcm.remote          p:INFO  > RemoteCmdRunner [scylla-test@10.142.0.65]: Command "sudo yum update scylla\* -y"       finished with status 0

< t:2020-01-01 22:47:16,535 f:remote.py       l:237  c:sdcm.remote          p:DEBUG > RemoteCmdRunner [scylla-test@10.142.0.65]: Running command "sudo systemctl start       scylla-server.service"...

## Failed to start scylla-jmx after upgraded db node 4

2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | scylla: [shard 0] init - starting storage service
2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | systemd: Started Scylla Server.
2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | sudo: pam_unix(sudo:session): session closed for user root
2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | systemd: Started Scylla JMX.
2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | systemd: Started Run Scylla Housekeeping restart mode.
2020-01-01T22:47:20 db-node-b642da38-0-4 !INFO    | systemd: Started Run Scylla Housekeeping daily mode.
2020-01-01T22:47:20 db-node-b642da38-0-4 !ERR     | systemd: Failed at step EXEC spawning /usr/lib/scylla/jmx/scylla-jmx: No such file or directory
2020-01-01T22:47:20 db-node-b642da38-0-4 !NOTICE  | systemd: scylla-jmx.service: main process exited, code=exited, status=203/EXEC
2020-01-01T22:47:20 db-node-b642da38-0-4 !NOTICE  | systemd: Unit scylla-jmx.service entered failed state.
2020-01-01T22:47:20 db-node-b642da38-0-4 !WARNING | systemd: scylla-jmx.service failed.

2020-01-01T22:46:43 db-node-b642da38-0-4 !INFO    | scylla: [shard 0] init - Scylla version 3.1.2-0.20191127.9d78d848e shutdown complete.
2020-01-01T22:47:16 db-node-b642da38-0-4 !INFO    | scylla: Scylla version 3.2.rc4-0.20200101.c8d5738a482 starting ...
2020-01-01T22:47:17 db-node-b642da38-0-4 !INFO    | scylla: [shard 0] init - Scylla version 3.2.rc4-0.20200101.c8d5738a482 starting.
2020-01-01T22:48:02 db-node-b642da38-0-4 !INFO    | scylla: [shard 0] init - Scylla version 3.2.rc4-0.20200101.c8d5738a482 initialization completed.
```

/CC @syuu1228

**Comment by roydahan** [comment]
@slivne this is a blocker for 3.2 (rolling upgrade)

**Comment by syuu1228** [comment]
@amoskong @roydahan  Is this after merged https://github.com/scylladb/scylla/commit/de5c06414b3a21962b4ea901a1f8f9fd290cd17b or before merged?

**Comment by amoskong** [comment]
> @amoskong @roydahan Is this after merged [scylladb/scylla@de5c064](https://github.com/scylladb/scylla/commit/de5c06414b3a21962b4ea901a1f8f9fd290cd17b) or before merged?

The issue started to happen after above commit was merged to branch-3.2

**Comment by syuu1228** [comment]
Okay.. will check

**Comment by syuu1228** [comment]
Looks like it happens because our .rpm specifies "%config(noreplace)" on /etc/sysconfig/scylla-jmx, scylla-jmx.service executes old verion of sysconfig file that specifies old path of scylla-jmx script:
``Failed at step EXEC spawning /usr/lib/scylla/jmx/scylla-jmx: No such file or directory``
It says /usr/lib/scylla but it should /opt/scylladb on 3.2.
I think safer approach to fix this is create symlink on /usr/lib/scylla/jmx so I'll send a patch to fix it, but I guess we can discuss we should keep "(noreplace)" option or not.

Note: why it only happens after merged https://github.com/scylladb/scylla/commit/de5c06414b3a21962b4ea901a1f8f9fd290cd17b is because we dropped /usr/lib/scylla symlink, now need to create symlink on /usr/lib/scylla for each file to keep compatibility with old script/config, like this issue.

**Comment by syuu1228** [comment]
Well previous comment was probably wrong, on some part.
Even if sysconfig file was not replaced with new one, it only provides following parameter:
``SCYLLA_JMX_LOCAL="-l /opt/scylladb/jmx"``
It doesn't contains scylla-jmx script path, it is hardcorded on systemd unit file and it's not specified as "(noreplace)" config file.
So after upgraded to 3.2, it should not execute /usr/lib/scylla/jmx/scylla-jmx even if sysconfig file is not upgraded. It may keep old paramter (-l /usr/lib/scylla/jmx) though.

**Comment by syuu1228** [comment]
It's more likely "systemctl --daemon-reload" has not executed, but I suppose .rpm should handled when upgraded.

**Comment by amoskong** [comment]
We only manually execute  "systemctl --daemon-reload" in rollback (remove
all packages and reinstall again, not smooth upgrade).

I just tested manually, execute  "systemctl --daemon-reload" can solve the
problem. But we need to consider if it's a graceful way.

Currently our upgrade documentation doesn't request users to
execute  "systemctl --daemon-reload" after upgrade.

what's your opinion?  @Shlomi Livne <shlomi@scylladb.com> @Tzach Livyatan
<tzach@scylladb.com>

**Comment by syuu1228** [comment]
We have a macro "%systemd_postun" on our .rpm, it should call "systemctl --daemon-reload" when it uninstalled (or upgraded), so if it working properly we don't need to call it manually, but it looks like not working..

**Comment by amoskong** [comment]
> We have a macro "%systemd_postun" on our .rpm, it should call "systemctl --daemon-reload" when it uninstalled (or upgraded), so if it working properly we don't need to call it manually, but it looks like not working..

Thanks.

**Comment by syuu1228** [comment]
Maybe we can just call "systemctl --daemon-reload" on .rpm scriptlet after installed (%post), not when uninstalled.

**Comment by syuu1228** [comment]
I'll send a pull-req to call daemon-reload on %post.

**Comment by syuu1228** [comment]
We might have same problem on scylla-server < 3.2, since we don't call daemon-reload on %post:
https://github.com/scylladb/scylla/blob/branch-3.1/dist/redhat/scylla.spec.mustache#L75
On 3.2 and master we call it, so should be no problem:
https://github.com/scylladb/scylla/blob/branch-3.2/scylla_post_install.sh#L72

**Comment by amoskong** [comment]
> We might have same problem on scylla-server < 3.2, since we don't call daemon-reload on %post:
> https://github.com/scylladb/scylla/blob/branch-3.1/dist/redhat/scylla.spec.mustache#L75
> On 3.2 and master we call it, so should be no problem:
> https://github.com/scylladb/scylla/blob/branch-3.2/scylla_post_install.sh#L72

systemd config of scylla-server or scylla-jmx wasn't change in old release, or the original config still works after upgrade or rollback. That's why we didn't touch real problem in original testing.

Let's see if the problem exists in rollback to 3.1  from 3.2 when your fix is backported to 3.2
